File: 2026/03- March/15-Sunday/solution.java
import java.util.*; 

class Fancy {

    private static final long MOD = 1_000_000_007;
    private List<Long> seq = new ArrayList<>();
    private long mul = 1;
    private long add = 0;

    public Fancy() {}

    public void append(int val) {
        long inv = modPow(mul, MOD-2);
        long v = (val - add) % MOD;
        if(v < 0) v += MOD;
        v = (v * inv) % MOD;
        seq.add(v);
    }

    public void addAll(int inc) {
        add = (add + inc) % MOD;
    }

    public void multAll(int m) {
        mul = (mul * m) % MOD;
        add = (add * m) % MOD;
    }

    public int getIndex(int idx) {
        if(idx >= seq.size()) return -1;
        long val = seq.get(idx);
        return (int)((val * mul + add) % MOD);
    }

    private long modPow(long base, long exp){
        long res = 1;
        base %= MOD;
        while(exp > 0){
            if((exp & 1) == 1)
                res = (res * base) % MOD;
            base = (base * base) % MOD;
            exp >>= 1;
        }
        return res;
    }
}

File: 2026/03- March/15-Sunday/solution.kt
class Fancy {

    private val MOD = 1_000_000_007L
    private val seq = ArrayList<Long>()
    private var mul = 1L
    private var add = 0L

    fun append(value: Int) {
        val inv = modPow(mul, MOD-2) 
        var v = (value - add) % MOD
        if (v < 0) v += MOD
        v = (v * inv) % MOD
        seq.add(v)
    }

    fun addAll(inc: Int) {
        add = (add + inc) % MOD
    }

    fun multAll(m: Int) {
        mul = (mul * m) % MOD
        add = (add * m) % MOD
    }

    fun getIndex(idx: Int): Int {
        if (idx >= seq.size) return -1
        val v = seq[idx]
        return ((v * mul + add) % MOD).toInt()
    }

    private fun modPow(base: Long, exp: Long): Long {
        var b = base % MOD
        var e = exp
        var res = 1L
        while (e > 0) {
            if (e and 1L == 1L) res = (res * b) % MOD
            b = (b * b) % MOD
            e = e shr 1
        }
        return res
    }
}
